extractor/src/main/java/org/schabi/newpipe/extractor/services/soundcloud/extractors/SoundcloudStreamExtractor.java

```diff
@@ -181,62 +186,144 @@ public String getHlsUrl() {
+                            .equals("progressive")) {
+                presence = true;
+                break;
+            }
+        }
+        return presence;
+    }
+
+    @Nonnull
+    private static String getTranscodingUrl(final String endpointUrl, final String protocol) throws IOException, ExtractionException {
+        final Downloader downloader = NewPipe.getDownloader();
+        final String apiStreamUrl = endpointUrl + "?client_id=" + SoundcloudParsingHelper.clientId();
+        final String response = downloader.get(apiStreamUrl).responseBody();
+        final JsonObject urlObject;
+        try {
+            urlObject = JsonParser.object().from(response);
+        } catch (final JsonParserException e) {
+            throw new ParsingException("Could not parse streamable url", e);
+        }
+        final String urlString = urlObject.getString("url");
+
+        if (protocol.equals("progressive")) {
+            return urlString;
+        } else if (protocol.equals("hls")) {
+            return getSingleUrlFromHlsManifest(urlString);
+        }
+        // else, unknown protocol
+        return "";
+    }
+
+    private static void setUpAudioStreams(final JsonArray transcodings,
+                                          final boolean mp3ProgressiveInStreams,
+                                          final List<AudioStream> audioStreams) {
+        for (final Object transcoding : transcodings) {
+            final JsonObject transcodingJsonObject = (JsonObject) transcoding;
+            final String mediaUrl;
+            final String preset = transcodingJsonObject.getString("preset");
+            final String protocol = transcodingJsonObject.getObject("format").getString("protocol");
+            final String url = transcodingJsonObject.getString("url");
+            final MediaFormat mediaFormat;
+            final int bitrate;
+
+            if (!isNullOrEmpty(url)) {
+                if (preset.contains("mp3")) {
+                    // Don't add the MP3 HLS stream if there is a progressive stream present
+                    // because the two have the same bitrate
+                    if (mp3ProgressiveInStreams && protocol.equals("hls")) {
+                        continue;
                     }
+                    mediaFormat = MediaFormat.MP3;
+                    bitrate = 128;
+                } else if (preset.contains("opus")) {
+                    mediaFormat = MediaFormat.OPUS;
+                    bitrate = 64;
+                } else {
+                    // Unknown format
+                    continue;
                 }
+
+                try {
+                    mediaUrl = getTranscodingUrl(url, protocol);
+                } catch (final Exception e) {
+                    // something went wrong when parsing this transcoding
+                    continue;
+                }
+                audioStreams.add(new AudioStream(mediaUrl, mediaFormat, bitrate));
             }
+        }
```

Comment:
I think this is easier to read:
```suggestion
        for (final Object transcoding : transcodings) {
            final JsonObject transcodingJsonObject = (JsonObject) transcoding;
            final String url = transcodingJsonObject.getString("url");

            if (isNullOrEmpty(url)) {
                continue;
            }

            final String mediaUrl;
            final String preset = transcodingJsonObject.getString("preset");
            final String protocol = transcodingJsonObject.getObject("format").getString("protocol");
            final MediaFormat mediaFormat;
            final int bitrate;

            if (preset.contains("mp3")) {
                // Don't add the MP3 HLS stream if there is a progressive stream present
                // because the two have the same bitrate
                if (mp3ProgressiveInStreams && protocol.equals("hls")) {
                    continue;
                }
                mediaFormat = MediaFormat.MP3;
                bitrate = 128;
            } else if (preset.contains("opus")) {
                mediaFormat = MediaFormat.OPUS;
                bitrate = 64;
            } else {
                // Unknown format
                continue;
            }

            try {
                mediaUrl = getTranscodingUrl(url, protocol);
                audioStreams.add(new AudioStream(mediaUrl, mediaFormat, bitrate));
            } catch (final Exception e) {
                // something went wrong when parsing this transcoding
            }

        }
```

Reply:
you could make this even shorter by removing the last `continue` and `else` branch and adding a check whether `mediaFormat` is not empty.